### ISSUE TITLE
Allow marking of plugins as optional

### ISSUE THREAD

**Issue by shalvah** [serverless/serverless#9100]
### Use case description

I often use some plugins that are only meant for a local environment. Specifically, [serverless-dotenv](https://www.npmjs.com/package/serverless-dotenv-plugin) and [serverless-offline](https://www.npmjs.com/package/serverless-offline). During deployment, I use `npm prune --production` to reduce the size of my deployment package by removing these dev dependencies. Unfortunately, this means my serverless deploy now fails, complaining that it can't find the plugins in my `serverless.yml`.

To fix this, my current workaround is to comment out these plugins and uncomment them on local. Very stressful and workflow-breaking.

### Proposed solution

Perhaps we could mark some plugins as optional?

```yaml
plugins:
- serverless-offline:
    optional: true
```

Or perhaps go for a more specific behaviour definition:

```yaml
plugins:
- serverless-offline:
    ignoreIfMissing: true
```

I believe this should be backwards -compatible with the current format.

**Comment by DavidWells** [comment]
+1 on the idea of having the plugin block accepting configuration directly where the plugin is defined.

Plugins often needed to specify their configuration in the `custom` block of serverless.yml. This can lead to confusion around what value relates to what plugin (especially if its a stack you didn't create)

```yml
custom:
  valueNeededForPlugin: 'xyz' # What is this for? ¯\_(ツ)_/¯
  webpack:
    webpackConfig: './webpack.js'  # this is a little more clear but could be defined far away from `plugins` block
    packager: 'npm'

# ... 
 
plugins:
  - my-serverless-plugin
  - serverless-webpack
```

Instead, consider the `plugins` key as an array of objects like so:

```yml
plugins:
  - my-serverless-plugin:
      valueNeededForPlugin: 'xyz' # Aha 💡  this value is used by the my-serverless-plugin
  - serverless-webpack:
      config: ./webpack.js
      packager: 'npm'
```

The benefits of this approach:

- Settings + plugin definition collocation makes it easier to grok what setting relates to what plugin
- It's backward compatible 🎉. If plugin array item = `string` use the same way, if plugin array item = `object`, use its config values
- Plugins could choose to adopt this format or continue using `custom` block if they choose.
- I think* this will make validating plugin input easier

**Comment by shalvah** [comment]
Agree. The `custom` approach also means each plugin has to pick its field name, and what happens if two plugins use the same field name?

One issue with the proposal, though, would be distinguishing between plugin-specified configs and framework-specified configs for a plugin. Do maybe something like:

```yaml
plugins:
- serverless-offline:
    ignoreIfMissing: true # <-- customise framework treatment of plugin
    config:
        port: 3040 # <-- customise plugin behaviour
```

**Comment by pgrzesik** [comment]
Hello @shalvah and @DavidWells - thank you both for very valuable input. We plan to revisit plugin support in the future and this is definitely worth considering. @medikoo What do you think of potentially extending `plugins` syntax to accept additional parameters?

**Comment by medikoo** [comment]
> During deployment, I use `npm prune --production` to reduce the size of my deployment package by removing these dev

Dev dependencies (so also plugins you've listed as ones) are by default not packaged to lambdas, so technically `npm prune --production` should make no difference here. What do I miss (?)

> Instead, consider the plugins key as an array of objects like so:

@DavidWells that's a very interesting idea! Definitely worth considering, it definitely makes sense as a replacement for `custom` block, as it's a bit different matter I've opened other issue for that at: https://github.com/serverless/serverless/issues/9116

> The custom approach also means each plugin has to pick its field name, and what happens if two plugins use the same field name?

Since we've introduced a config schema validation. Plugins are expected to provide a schema for properties their introduce, and defining schema for same property twice will crash with meaningful error
